tests/timeseries/test_quantities.py

```diff
@@ -527,3 +531,31 @@ def test_invalid_multiplications() -> None:
             _ = quantity * 200.0  # type: ignore
         with pytest.raises(TypeError):
             quantity *= 200.0  # type: ignore
+
+
+def test_to_and_from_string() -> None:
+    """Test string parsing and formatting."""
+    random.seed(0)
+    exponents_to_test = [0, 3, 6, 9]
+
+    for exponent in exponents_to_test:
+        for _ in range(100):
+            value = random.random() * 10**exponent
+
+            power_str = f"{Power.from_watts(value):.{exponent}}"
+            assert f"{Power.from_string(power_str):.{exponent}}" == power_str
+
+            voltage_str = f"{Voltage.from_volts(value):.{exponent}}"
+            assert f"{Voltage.from_string(voltage_str):.{exponent}}" == voltage_str
+
+            current_str = f"{Current.from_amperes(value):.{exponent}}"
+            assert f"{Current.from_string(current_str):.{exponent}}" == current_str
+
+            energy_str = f"{Energy.from_watt_hours(value):.{exponent}}"
+            assert f"{Energy.from_string(energy_str):.{exponent}}" == energy_str
+
+            freq_str = f"{Frequency.from_hertz(value):.{exponent}}"
+            assert f"{Frequency.from_string(freq_str):.{exponent}}" == freq_str
+
+            pct_str = f"{Percentage.from_fraction(value):.{exponent}}"
+            assert f"{Percentage.from_string(pct_str):.{exponent}}" == pct_str
```

Comment:
It looks like you are re-implementing [hypothesis](https://hypothesis.readthedocs.io/) :grimacing: 

What you are doing is basically the [first example](https://hypothesis.readthedocs.io/en/latest/quickstart.html#an-example) but you need to [generate inputs with floats](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.floats) instead.

Reply:
BTW, to include all types you can also parametrize it (untested):
```py
@pytest.parametrize("quantity_type", [Power, Voltage, ...])
@pytest.parametrize("exponent", [0, 3, 6, 9])
@hypothesis.given(value=st.floats(min_value=-1.0, max_value=1.0))
def test_to_and_from_string(
        quantity_type: type[Quantity],
        exponent: int,
        value: float
    ) -> None:
    """Test string parsing and formatting."""
    quantity = quantity_type.__new__(quantity_type)
    quantity._base_value = value * 10**exponent
    quantity_str = f"{quantity:.{exponent}}"
    assert f"{quantity_type.from_string(quantity_str):.{exponent}}" == quantity_str
```

The construction is a bit hacky, but OK IMHO for a unit test. You could also parametrize on tuples `"quantity_ctor,  quantity_from_str", [(Power.from_watts, Power.from_string), (...), ...]` if you want a less hacky but more verbose approach.